core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -0,0 +1,568 @@
+import io.trino.sql.planner.plan.AggregationNode.Aggregation;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.tree.Cast;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.NodeRef;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SearchedCaseExpression;
+import io.trino.sql.tree.SymbolReference;
+import io.trino.sql.tree.WhenClause;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.SystemSessionProperties.isPreAggregateCaseAggregationsEnabled;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static io.trino.sql.analyzer.TypeSignatureTranslator.toSqlType;
+import static io.trino.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static io.trino.sql.planner.plan.AggregationNode.singleGroupingSet;
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+import static java.util.function.Predicate.not;
+
+/**
+ * Rule that transforms selected aggregations from:
+ * <pre> {@code
+ * - Aggregation[key]: sum(aggr1), sum(aggr2), ..., sum(aggrN)
+ *   - Project:
+ *       aggr1 = CASE WHEN col=1 THEN expr ELSE 0
+ *       aggr2 = CASE WHEN col=2 THEN expr ELSE null
+ *       ...
+ *       aggrN = CASE WHEN col=N THEN expr
+ *     - source
+ * }
+ * </pre>
+ * into:
+ * <pre> {@code
+ * - Aggregation[key]: sum(aggr1), sum(aggr2), ..., sum(aggrN)
+ *   - Project:
+ *       aggr1 = CASE WHEN col=1 THEN pre_aggregate
+ *       aggr2 = CASE WHEN col=2 THEN pre_aggregate
+ *       ..
+ *       aggrN = CASE WHEN col=N THEN pre_aggregate
+ *     - Aggregation[key, col]: pre_aggregate = sum(expr)
+ *       - source
+ * }
+ * </pre>
+ */
+public class PreAggregateCaseAggregations
+        implements Rule<AggregationNode>
+{
+    private static final int MIN_AGGREGATION_COUNT = 4;
+    private static final Set<String> ALLOWED_FUNCTIONS = ImmutableSet.of("max", "min", "sum");
+    private static final Capture<ProjectNode> PROJECT_CAPTURE = newCapture();
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
```

Comment:
Should we check that there is single grouping set? What if there is an empty grouping set?

Reply:
Multiple grouping sets are not supported (see `testDoesNotFireWithGroupingSets`). I think it would still work, but we don't recognize such pattern (with `GroupIdNode`).

Default aggregation are supported and there are tests for it